File: python/processor.py
"""
Contoso Toyland - Order Processor (Python)
Processes orders from the data/orders.csv file and logs results.
"""
import csv
import logging
import os
import sys
import traceback

# Ensure logs directory exists
os.makedirs("logs", exist_ok=True)

# Configure "noisy" logging to simulate real production environment
logging.basicConfig(
    filename='logs/python.log',
    filemode='w',  # Overwrite each run for demo clarity
    level=logging.DEBUG,
    format='%(asctime)s - %(levelname)s - [OrderProcessor] - %(message)s'
)

logger = logging.getLogger(__name__)


def validate_inventory(product_id):
    """Simulates inventory check - adds noise to logs."""
    logger.debug(f"Checking inventory for product: {product_id}")
    logger.debug(f"Inventory check passed for {product_id}")
    return True


def calculate_shipping(quantity):
    """Simulates shipping calculation - adds noise to logs."""
    logger.debug(f"Calculating shipping for {quantity} items")
    base_rate = 5.99
    per_item = 1.50
    shipping = base_rate + (quantity * per_item)
    logger.debug(f"Shipping calculated: ${shipping:.2f}")
    return shipping


# Global discount cache - simulates a caching layer
_discount_cache = {}

def get_discount_tier(product_name):
    """
    Returns discount tier based on product category.
    BUG: Accessing dictionary key that doesn't exist!
    """
    tiers = {
        "toy": 0.10,
        "game": 0.15,
        "puzzle": 0.12
    }
    # BUG: product_name.lower() might not match any key - returns None then used in math
    category = product_name.lower().split()[0] if product_name else "unknown"
    logger.debug(f"product_name = '{product_name}', extracted category = '{category}'")
    result = tiers.get(category)  # Returns None if not found - causes TypeError later!
    logger.debug(f"tiers.get('{category}') = {result}")
    return result


def apply_holiday_discount(total, order_id, product_name=None):
    """
    Applies the holiday discount for the January Rush sale.
    BUG: Multiple issues - None handling, division, caching!
    """
    logger.info(f"Applying holiday discount for order {order_id}")
    
    # BUG 1: get_discount_tier can return None
    tier_discount = get_discount_tier(product_name)
    logger.debug(f"tier_discount = {tier_discount}")  # Will show None for unknown categories!
    base_rate = 0.15
    
    # BUG 2: If tier_discount is None, this causes TypeError: unsupported operand type(s)
    logger.debug(f"Attempting: discount_rate = {base_rate} + {tier_discount}")
    discount_rate = base_rate + tier_discount
    
    # BUG 3: Division by zero when total is 0
    logger.debug(f"total = {total}, order_id in cache = {order_id in _discount_cache}")
    if order_id in _discount_cache:
        # Simulate cache hit ratio calculation
        logger.debug(f"Attempting: cache_bonus = {_discount_cache[order_id]} / {total}")
        cache_bonus = _discount_cache[order_id] / total  # ZeroDivisionError if total=0
        discount_rate += cache_bonus
    
    # BUG 4: No cap on discount_rate - could exceed 100% causing negative prices!
    logger.debug(f"discount_rate = {discount_rate}")
    discount_amount = total * discount_rate
    final_price = total - discount_amount
    
    # Cache this discount for "analytics"
    _discount_cache[order_id] = discount_amount
    
    logger.debug(f"Discount applied: ${discount_amount:.2f} off, new total: ${final_price:.2f}")
    return final_price


def process_order(order):
    """
    Main order processing function.
    Processes a single order and returns success/failure status.
    """
    order_id = order.get('order_id', 'UNKNOWN')
    
    try:
        logger.info(f"========== Processing Order {order_id} ==========")
        logger.info(f"Customer order received: {order['product_name']}")
        
        # Step 1: Parse and validate quantity
        logger.debug(f"Parsing quantity value: '{order['quantity']}'")
        qty = int(order['quantity'])
        
        # Step 2: Parse and validate price
        logger.debug(f"Parsing unit price value: '{order['unit_price']}'")
        price = float(order['unit_price'])
        
        # Step 3: Business rule validation
        if qty < 0:
            raise ValueError(f"Quantity cannot be negative. Received: {qty}")
        
        if qty == 0:
            logger.warning(f"Order {order_id} has zero quantity - flagging for review")
        
        # Step 4: Inventory check
        validate_inventory(order['product_id'])
        
        # Step 5: Calculate totals
        subtotal = qty * price
        logger.debug(f"Subtotal calculated: {qty} x ${price:.2f} = ${subtotal:.2f}")
        
        # Step 6: Apply holiday discount
        discounted_total = apply_holiday_discount(subtotal, order_id, order['product_name'])
        
        # Step 7: Calculate tax
        tax_rate = 0.08
        tax = discounted_total * tax_rate
        logger.debug(f"Tax calculated: ${tax:.2f}")
        
        # Step 8: Calculate shipping
        shipping = calculate_shipping(qty)
        
        # Step 9: Final total
        final_total = discounted_total + tax + shipping
        
        logger.info(f"Order {order_id} processed successfully!")
        logger.info(f"  Product: {order['product_name']}")
        logger.info(f"  Subtotal: ${subtotal:.2f}")
        logger.info(f"  After Discount: ${discounted_total:.2f}")
        logger.info(f"  Tax: ${tax:.2f}")
        logger.info(f"  Shipping: ${shipping:.2f}")
        logger.info(f"  FINAL TOTAL: ${final_total:.2f}")
        logger.info(f"========== Order {order_id} Complete ==========\n")
        
        return True

    except ValueError as ve:
        logger.error(f"VALIDATION ERROR on order {order_id}: {str(ve)}")
        logger.error(f"  Raw data - Qty: '{order.get('quantity')}', Price: '{order.get('unit_price')}'")
        logger.debug(f"Stack trace: {traceback.format_exc()}")
        return False
        
    except Exception as e:
        logger.critical(f"CRITICAL SYSTEM FAILURE processing order {order_id}")
        logger.critical(f"  Error Type: {type(e).__name__}")
        logger.critical(f"  Error Message: {str(e)}")
        logger.critical(f"  Full Stack Trace:\n{traceback.format_exc()}")
        return False


def main():
    """Main entry point for the order processor."""
    # Resolve path relative to this script's location
    script_dir = os.path.dirname(os.path.abspath(__file__))
    csv_path = os.path.join(script_dir, '..', 'data', 'orders.csv')
    
    print("=" * 60)
    print("  CONTOSO TOYLAND - Order Processing System (Python)")
    print("  Holiday Rush Batch Processor v2.1")
    print("=" * 60)
    
    logger.info("=" * 60)
    logger.info("CONTOSO TOYLAND ORDER PROCESSOR STARTED")
    logger.info(f"Processing file: {csv_path}")
    logger.info("=" * 60)
    
    if not os.path.exists(csv_path):
        error_msg = f"FATAL: Data file not found at {csv_path}"
        print(error_msg)
        logger.critical(error_msg)
        return

    success_count = 0
    error_count = 0
    total_orders = 0

    with open(csv_path, mode='r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            total_orders += 1
            if process_order(row):
                success_count += 1
            else:
                error_count += 1
    
    # Summary
    logger.info("=" * 60)
    logger.info("BATCH PROCESSING COMPLETE")
    logger.info(f"  Total Orders: {total_orders}")
    logger.info(f"  Successful: {success_count}")
    logger.info(f"  Failed: {error_count}")
    logger.info("=" * 60)
    
    print(f"\nProcessing complete!")
    print(f"  Total Orders: {total_orders}")
    print(f"  Successful:   {success_count}")
    print(f"  Failed:       {error_count}")
    print(f"\nCheck logs/python.log for detailed output.")


if __name__ == "__main__":
    main()

File: node/processor.js
/**
 * Contoso Toyland - Order Processor (Node.js)
 * Processes orders from the data/orders.csv file and logs results.
 */
const fs = require('fs');
const path = require('path');

// Ensure logs directory exists
const logDir = path.join(__dirname, '..', 'logs');
if (!fs.existsSync(logDir)) {
    fs.mkdirSync(logDir, { recursive: true });
}

const logFile = path.join(logDir, 'node.log');

// Clear log file at start
fs.writeFileSync(logFile, '');

/**
 * Writes a formatted log entry to the log file
 */
function log(level, message) {
    const timestamp = new Date().toISOString();
    const logEntry = `${timestamp} - ${level.padEnd(8)} - [OrderProcessor] - ${message}\n`;
    fs.appendFileSync(logFile, logEntry);
}

/**
 * Simulates inventory check - adds noise to logs
 */
function validateInventory(productId) {
    log('DEBUG', `Checking inventory for product: ${productId}`);
    log('DEBUG', `Inventory check passed for ${productId}`);
    return true;
}

/**
 * Simulates shipping calculation - adds noise to logs
 */
function calculateShipping(quantity) {
    log('DEBUG', `Calculating shipping for ${quantity} items`);
    const baseRate = 5.99;
    const perItem = 1.50;
    const shipping = baseRate + (quantity * perItem);
    log('DEBUG', `Shipping calculated: $${shipping.toFixed(2)}`);
    return shipping;
}

// Global state - simulates external config
let discountConfig = null;
let orderHistory = [];

// Track load attempts to make bug deterministic
let configLoadAttempts = 0;

/**
 * Loads discount configuration from "external source"
 * BUG: Race condition - config might not be loaded!
 */
function loadDiscountConfig() {
    configLoadAttempts++;
    
    // BUG: Config only loads on 3rd+ attempt - first orders always fail!
    // Simulates async config that isn't ready immediately
    if (configLoadAttempts >= 3) {
        discountConfig = {
            baseRate: 0.15,
            bonusCategories: ['RC', 'Robot'],
            maxDiscount: 0.25
        };
    }
    // BUG: First 2 orders will have null discountConfig!
}

/**
 * Gets bonus rate based on product ID
 * BUG: Array index out of bounds!
 */
function getBonusRate(productId) {
    const parts = productId.split('-');
    log('DEBUG', `productId = '${productId}', parts = ${JSON.stringify(parts)}`);
    // BUG: Assumes parts[1] exists - crashes on malformed IDs
    log('DEBUG', `Attempting to access parts[1]: ${parts[1]}`);
    const category = parts[1].toUpperCase();
    
    // BUG: Accessing property of potentially null discountConfig
    if (discountConfig.bonusCategories.includes(category)) {
        return 0.05;
    }
    return 0;
}

/**
 * Calculates loyalty bonus from order history
 * BUG: Off-by-one error and undefined access!
 */
function getLoyaltyBonus(orderId) {
    orderHistory.push(orderId);
    
    // BUG: Off-by-one - accesses index that doesn't exist yet
    log('DEBUG', `orderHistory.length = ${orderHistory.length}, accessing index = ${orderHistory.length}`);
    const previousOrder = orderHistory[orderHistory.length];
    log('DEBUG', `previousOrder = ${previousOrder}`);  // Will always be undefined!
    
    // BUG: Calling method on undefined
    if (previousOrder && previousOrder.startsWith('CT-100')) {
        return 0.02;
    }
    return 0;
}

/**
 * Applies the holiday discount for the January Rush sale.
 * BUG: Multiple issues - null refs, array bounds, race conditions!
 */
function applyHolidayDiscount(total, orderId, productId) {
    log('INFO', `Applying holiday discount for order ${orderId}`);
    
    // Load config (might fail silently)
    loadDiscountConfig();
    
    // BUG 1: discountConfig might be null here
    log('DEBUG', `discountConfig = ${JSON.stringify(discountConfig)}`);
    log('DEBUG', `configLoadAttempts = ${configLoadAttempts}`);
    let discountRate = discountConfig.baseRate;
    
    // BUG 2: getBonusRate has array index issues
    discountRate += getBonusRate(productId);
    
    // BUG 3: getLoyaltyBonus has off-by-one error
    discountRate += getLoyaltyBonus(orderId);
    
    // BUG 4: No cap check - discount could exceed 100%
    log('DEBUG', `discountRate = ${discountRate}`);
    const discountAmount = total * discountRate;
    const finalPrice = total - discountAmount;
    
    log('DEBUG', `Discount applied: $${discountAmount.toFixed(2)} off, new total: $${finalPrice.toFixed(2)}`);
    return finalPrice;
}

/**
 * Main order processing function.
 * Processes a single order and returns success/failure status.
 */
function processOrder(order) {
    const orderId = order.order_id || 'UNKNOWN';
    
    try {
        log('INFO', `========== Processing Order ${orderId} ==========`);
        log('INFO', `Customer order received: ${order.product_name}`);
        
        // Step 1: Parse and validate quantity
        log('DEBUG', `Parsing quantity value: '${order.quantity}'`);
        const qty = parseInt(order.quantity, 10);
        
        // Step 2: Parse and validate price
        log('DEBUG', `Parsing unit price value: '${order.unit_price}'`);
        const price = parseFloat(order.unit_price);
        
        // Check for parsing failures
        if (isNaN(qty)) {
            throw new Error(`Invalid quantity value: '${order.quantity}' is not a number`);
        }
        
        if (isNaN(price)) {
            throw new Error(`Invalid price value: '${order.unit_price}' is not a number`);
        }
        
        // Step 3: Business rule validation
        if (qty < 0) {
            throw new Error(`Quantity cannot be negative. Received: ${qty}`);
        }
        
        if (qty === 0) {
            log('WARNING', `Order ${orderId} has zero quantity - flagging for review`);
        }
        
        // Step 4: Inventory check
        validateInventory(order.product_id);
        
        // Step 5: Calculate totals
        const subtotal = qty * price;
        log('DEBUG', `Subtotal calculated: ${qty} x $${price.toFixed(2)} = $${subtotal.toFixed(2)}`);
        
        // Step 6: Apply holiday discount
        const discountedTotal = applyHolidayDiscount(subtotal, orderId, order.product_id);
        
        // Step 7: Calculate tax
        const taxRate = 0.08;
        const tax = discountedTotal * taxRate;
        log('DEBUG', `Tax calculated: $${tax.toFixed(2)}`);
        
        // Step 8: Calculate shipping
        const shipping = calculateShipping(qty);
        
        // Step 9: Final total
        const finalTotal = discountedTotal + tax + shipping;
        
        log('INFO', `Order ${orderId} processed successfully!`);
        log('INFO', `  Product: ${order.product_name}`);
        log('INFO', `  Subtotal: $${subtotal.toFixed(2)}`);
        log('INFO', `  After Discount: $${discountedTotal.toFixed(2)}`);
        log('INFO', `  Tax: $${tax.toFixed(2)}`);
        log('INFO', `  Shipping: $${shipping.toFixed(2)}`);
        log('INFO', `  FINAL TOTAL: $${finalTotal.toFixed(2)}`);
        log('INFO', `========== Order ${orderId} Complete ==========\n`);
        
        return true;

    } catch (error) {
        log('ERROR', `VALIDATION ERROR on order ${orderId}: ${error.message}`);
        log('ERROR', `  Raw data - Qty: '${order.quantity}', Price: '${order.unit_price}'`);
        
        // Log stack trace for debugging
        if (error.message.includes('not a number')) {
            log('CRITICAL', `DATA CORRUPTION detected in order ${orderId}`);
            log('CRITICAL', `  This indicates malformed CSV data or upstream system failure`);
            log('CRITICAL', `  Stack trace: ${error.stack}`);
        }
        
        return false;
    }
}

/**
 * Parses CSV content into array of objects
 */
function parseCSV(content) {
    const lines = content.trim().split('\n');
    const headers = lines[0].split(',').map(h => h.trim());
    const records = [];
    
    for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(',');
        const record = {};
        headers.forEach((header, index) => {
            record[header] = values[index] ? values[index].trim() : '';
        });
        records.push(record);
    }
    
    return records;
}

/**
 * Main entry point for the order processor
 */
function main() {
    const csvPath = path.join(__dirname, '..', 'data', 'orders.csv');
    
    console.log('='.repeat(60));
    console.log('  CONTOSO TOYLAND - Order Processing System (Node.js)');
    console.log('  Holiday Rush Batch Processor v2.1');
    console.log('='.repeat(60));
    
    log('INFO', '='.repeat(60));
    log('INFO', 'CONTOSO TOYLAND ORDER PROCESSOR STARTED');
    log('INFO', `Processing file: ${csvPath}`);
    log('INFO', '='.repeat(60));
    
    if (!fs.existsSync(csvPath)) {
        const errorMsg = `FATAL: Data file not found at ${csvPath}`;
        console.error(errorMsg);
        log('CRITICAL', errorMsg);
        return;
    }

    const content = fs.readFileSync(csvPath, 'utf8');
    const orders = parseCSV(content);
    
    let successCount = 0;
    let errorCount = 0;
    const totalOrders = orders.length;

    for (const order of orders) {
        if (processOrder(order)) {
            successCount++;
        } else {
            errorCount++;
        }
    }
    
    // Summary
    log('INFO', '='.repeat(60));
    log('INFO', 'BATCH PROCESSING COMPLETE');
    log('INFO', `  Total Orders: ${totalOrders}`);
    log('INFO', `  Successful: ${successCount}`);
    log('INFO', `  Failed: ${errorCount}`);
    log('INFO', '='.repeat(60));
    
    console.log('\nProcessing complete!');
    console.log(`  Total Orders: ${totalOrders}`);
    console.log(`  Successful:   ${successCount}`);
    console.log(`  Failed:       ${errorCount}`);
    console.log('\nCheck logs/node.log for detailed output.');
}

main();
